xmpp-vala/src/module/xep/0047_in_band_bytestreams.vala

```diff
@@ -0,0 +1,437 @@
+        }
+        public override bool close(Cancellable? cancellable = null) throws IOError {
+            return connection.close_write(cancellable);
+        }
+        public override async bool close_async(int io_priority = GLib.Priority.DEFAULT, Cancellable? cancellable = null) throws IOError {
+            return yield connection.close_write_async(io_priority, cancellable);
+        }
+    }
+
+    private Input input;
+    private Output output;
+    public override InputStream input_stream { get { return input; } }
+    public override OutputStream output_stream { get { return output; } }
+
+    public enum State {
+        WAITING_FOR_CONNECT,
+        CONNECTING,
+        CONNECTED,
+        DISCONNECTING,
+        DISCONNECTED,
+        ERROR,
+    }
+    public State state { get; private set; }
+    Jid receiver_full_jid;
+    public string sid { get; private set; }
+    int block_size;
+    int local_seq = 0;
+    int remote_ack = 0;
+    internal int remote_seq = 0;
+
+    bool input_closed = false;
+    bool output_closed = false;
+
+    // ERROR
+    string? error = null;
+
+    XmppStream stream;
+
+    SourceFunc? read_callback = null;
+    SourceFunc? write_callback = null;
+    // Need `Bytes` instead of `uint8[]` because the latter doesn't work in
+    // parameter position of `LinkedList`.
+    LinkedList<Bytes> received = new LinkedList<Bytes>();
+
+    private Connection(XmppStream stream, Jid receiver_full_jid, string sid, int block_size, bool initiate) {
+        this.stream = stream;
+        this.receiver_full_jid = receiver_full_jid;
+        this.sid = sid;
+        this.block_size = block_size;
+        this.state = initiate ? State.CONNECTING : State.WAITING_FOR_CONNECT;
+
+        input = new Input(this);
+        output = new Output(this);
+    }
+
+    public void set_read_calllback(SourceFunc callback) throws IOError {
+        if (read_callback != null) {
+            throw new IOError.PENDING("only one async read is permitted at a time on an in-band bytestream");
+        }
+        read_callback = callback;
+    }
+    public void set_write_calllback(SourceFunc callback) throws IOError {
+        if (write_callback != null) {
+            throw new IOError.PENDING("only one async write is permitted at a time on an in-band bytestream");
+        }
+        write_callback = callback;
+    }
+    public void trigger_read_callback() {
+        if (read_callback != null) {
+            Idle.add((owned) read_callback);
+            read_callback = null;
+        }
+    }
+    public void trigger_write_callback() {
+        if (write_callback != null) {
+            Idle.add((owned) write_callback);
+            write_callback = null;
+        }
+    }
+
+    public async ssize_t read_async(uint8[]? buffer, int io_priority = GLib.Priority.DEFAULT, Cancellable? cancellable = null) throws IOError {
+        // TODO(hrxi): cancellable?
```

Comment:
- Check for `cancellable.is_cancelled()` in the loop and early return if set.
- Use `cancellable.cancelled()` signal to trigger the read callback.

Such that future calls to `cancellable.cancel()` will cancel the execution of `read_async()` as soon as the main loop gets free. Note that I don't think we need this behavior anywhere and likely will just pass null to the cancellable, but it shouldn't be hard to implement either and it might be considered functional at some point.

**Not** blocking for merging.

Reply:
Done.